code/modules/crafting/items.dm

```diff
@@ -979,7 +979,253 @@
 				/obj/item/weldingtool/advanced,
 				/obj/item/stock_parts/cell/ammo/mfc,
 				/obj/item/stock_parts/cell/ammo/ec,
+				/obj/item/stock_parts/cell/ammo/ecp,
 				/obj/item/megaphone)
 
+/obj/item/experimental
+	name = "Experimental component"
+	desc = "What could this do..."
+	icon = 'icons/obj/assemblies.dmi'
+	icon_state = "radio-multitool"
+
+/obj/item/experimental/attackby(obj/item/W, mob/user, params)
+	. = ..()
+	if(istype(W, /obj/item/gun/ballistic/shotgun))
+		to_chat(usr, "You can't improve [W.name]...")
+		return
+	if(istype(W, /obj/item/gun/ballistic))
+		gun(W, user)
```

Comment:
NO. NO NO NO NO NO.
You should create ONE tinker proc on `/obj/item`, and override it on the individual types.